src/Owin.Security.Providers.CanvasLMS/CanvasAuthenticationHandler.cs

```diff
@@ -181,7 +131,60 @@ protected override Task ApplyResponseChallengeAsync()
 
         public override async Task<bool> InvokeAsync()
         {
-            return await InvokeReplyPathAsync();
+            return await InvokeReplyPathAsync() || await RefreshAccessTokenAsync();
+        }
+
+        private async Task<bool> RefreshAccessTokenAsync()
+        {
+            var accessTokenExpiration = ParseExpiration();
+            if (accessTokenExpiration > DateTimeOffset.Now.AddMinutes(1))
+                return false;
+
+            var refreshToken = Context.Authentication.User?.FindFirst(Constants.CanvasRefreshToken)?.Value;
+            if (string.IsNullOrEmpty(refreshToken))
+                return false;
+
+            _logger.WriteInformation("Requesting new access token.");
+
+            try
+            {
+                var response = await RequestToken("refresh_token", refreshToken);
+                var identity = Context.Authentication.User.Identities
+                    .FirstOrDefault(ci => ci.AuthenticationType == Options.SignInAsAuthenticationType);
+
+                var context = await Authenticate(response, refreshToken: refreshToken);
+
+                if (identity == null)
+                {
+                    identity = context.Identity;
+                }
+                else
+                {
+                    var claims = identity.Claims.ToList();
+                    foreach (var c in claims)
+                        identity.RemoveClaim(c);
```

Comment:
Correct. Would be super handy, as this implementation is O(n²).

Need to capture into temp `ToList()` to avoid missing claims while changing the underlying collections. (No "Collection was modified; enumeration may not execute" error due to internal use of an iterator.)